activesupport/lib/active_support/core_ext/range/each.rb

```diff
@@ -11,13 +11,45 @@ def each(&block)
 
     def step(n = 1, &block)
       ensure_iteration_allowed
-      super
+
+      case n
+      when ActiveSupport::Duration
+        duration_step(n, &block)
+      else
+        super
+      end
     end
 
     private
       def ensure_iteration_allowed
+        raise TypeError, "can't iterate beginless range" unless self.begin
         raise TypeError, "can't iterate from #{first.class}" if first.is_a?(TimeWithZone)
       end
+
+      def duration_step(step, &block)
+        if block_given?
+          iterate(step, &block)
+
+          self
+        else
+          Enumerator.new do |yielder|
+            iterate(step) do |value|
+              yielder << value
+            end
+          end
+        end
+      end
+
+      def iterate(step, &block)
+        value, i = self.begin, 0
+
+        while cover?(value)
+          yield value
+
+          i += 1
+          value = self.begin + (step * i)
+        end
+      end
```

Comment:
This implementation was chosen to align [with Ruby implementation](https://github.com/ruby/ruby/blob/5124f9ac7513eb590c37717337c430cb93caa151/range.c#L510-L516), i.e. if first value is $B$, step is $S$, then all other values are of form $B + nS, n \ge 0$.

For dates this means that if e.g. first value is `2020-01-31` (end of month) and step is `1.month`, then the following values will be:
- `2020-02-29`
- `2020-03-31` (note that it's still end of month and not the 29th)
- `2020-04-30`
- etc.